Add handling of beatmap links on startup

peppy opened the pull request:
This allows clicking the "osu!direct" button on beatmap info pages when lazer is in a *non-open state* and having it open to the correct beatmap.

Note that this doesn't include the IPC implementation (required to handle the case where osu! is already running), which requires a bit more restructuring. Don't want to put in the effort for that just yet, and I believe this is one of the more common use cases (at least for me).

peppy committed:
Add a centralised constant for the osu URL schema protocol

peppy committed:
Add handling of beatmap links on startup

bdach commented on osu.Game/Online/Chat/MessageFormatter.cs:
```diff
@@ -269,10 +269,10 @@ private static MessageFormatterResult format(string toFormat, int startIndex = 0
 
             // handle editor times
-            handleMatches(time_regex, "{0}", "osu://edit/{0}", result, startIndex, LinkAction.OpenEditorTimestamp);
+            handleMatches(time_regex, "{0}", $"{OsuGameBase.OSU_PROTOCOL}edit/{0}", result, startIndex, LinkAction.OpenEditorTimestamp);
```
This needs more escaping (see test failures):

```suggestion
            handleMatches(time_regex, "{0}", $"{OsuGameBase.OSU_PROTOCOL}edit/{{0}}", result, startIndex, LinkAction.OpenEditorTimestamp);
```

bdach commented on osu.Game/Online/Chat/MessageFormatter.cs:
```diff
@@ -269,10 +269,10 @@ private static MessageFormatterResult format(string toFormat, int startIndex = 0
 
             // handle channels
-            handleMatches(channel_regex, "{0}", "osu://chan/{0}", result, startIndex, LinkAction.OpenChannel);
+            handleMatches(channel_regex, "{0}", $"{OsuGameBase.OSU_PROTOCOL}chan/{0}", result, startIndex, LinkAction.OpenChannel);
```
Same here:

```suggestion
            handleMatches(channel_regex, "{0}", $"{OsuGameBase.OSU_PROTOCOL}chan/{{0}}", result, startIndex, LinkAction.OpenChannel);
```

bdach commented on osu.Game.Tests/Visual/Navigation/TestSceneStartupBeatmapDisplay.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace osu.Game.Tests.Visual.Navigation
+{
+    public class TestSceneStartupBeatmapDisplay : OsuGameTestScene
```
I find it kinda weird that this test scene kind of just... doesn't work, either in debug or release config - sure, the overlay loads, but it's blank, because the `TestOsuGame` is using the provided dummy `API` so requests will never actually succeed.

Thoughts on mocking the requests to make this test actually show something?

bdach commented on osu.Game/OsuGame.cs:
```diff
@@ -898,8 +899,41 @@ private void handleStartupImport()
+            if (!url.StartsWith(OSU_PROTOCOL, StringComparison.Ordinal))
+                throw new ArgumentException("Invalid osu URL provided.", nameof(url));
+
+            string[] pieces = url.Split('/');
```
I initially wanted to propose using `StringSplitOptions.RemoveEmptyEntries` here, but I guess not specifying it makes the indexes below also implicitly check that the URL is formed correctly (i.e. with removing empty entries, `osu:///////s/1` would also be accepted). Maybe best to pre-empt that with an inline comment?

peppy replied:
I'm not really sure how to phrase the inline comment, but am open to adding one if you word it.

bdach replied:
I've pushed an inline comment in - please check whether you find it legible.

peppy committed:
Fix missing escaping causing test failures

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy committed:
Fix one more case of escaping not being present

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy committed:
Add test coverage of beatmap set overlay actually showing requested beatmap

peppy committed:
Fix current selection not being correctly maintained when `BeatmapPicker` updates its display

peppy commented:
I've added test coverage to this, and also fixed a case where the proposed beatmap is not the one actually displayed in the process (figured it's not worth moving to its own PR, but can if required).

bdach committed:
Add comment about split usage in osu:// protocol link handling

bdach committed:
Merge branch 'master' into startup-protocol-handling

Joehuu commented on osu.Game/OsuGame.cs:
```diff
@@ -898,8 +899,44 @@ private void handleStartupImport()
+            }
+        }
+
+        private void handleOsuProtocolUrl(string url)
```
Any reason to not use the existing `HandleLink(url)` method in `OsuGame`? That'll lead to the existing logic here:

https://github.com/ppy/osu/blob/d0d0830888159fcd92a110d84d2a27b5e8892680/osu.Game/Online/Chat/MessageFormatter.cs#L204-L243

bdach replied:
That whole structure `Link` / `LinkDetails` [was previously called out as pretty dodgy](https://github.com/ppy/osu/issues/11754) (and I agree), so maybe best not to? Not sure.

peppy replied:
I had a feeling this was already done somewhere, but couldn't immediately find the flow. I've switched to this as it seems pretty pointless to reimplement what we already have.

peppy committed:
Use existing `HandleLink` flow rather than reimplmenting

bdach commented:
I'll note that with the latest changes this will also show user profiles and join chat channels (although that last one is a bit weird because the channels will be joined, but the chat overlay won't be focused on launch).

bdach committed:
Merge branch 'master' into startup-protocol-handling
